### PR TITLE
[Core][aDAG] support multi readers in multi node when dag is created from an actor

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Currently, when a DAG is created from an actor, we are using different mechanism from a driver. In a driver we create a ProxyActor vs actor we are just using the actor itself.

This inconsistent mechanism is prone to error. As an example, I found when we support multi reader in multi node, we have deadlock because the driver actor needs to call `ray.get(a.allocate_channel.remote())` for a downstream actor while the downstream actor calls `ray.get(driver_actor.create_ref.remote())`. 

This fixes the issue by making ProxyActor as the default mechanism even when a dag is created inside an actor. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 1]
```diff
@@ -703,7 +713,7 @@ def _get_self_actor_id(self) -> str:
         use the actor ID of the DAGDriverProxyActor.
         """
         actor_id = ray.get_runtime_context().get_actor_id()
-        if actor_id is None:
+        if self._read_by_multi_output_node:
             # The reader is the driver process.
```
Comment (kevin85421) [inline python/ray/experimental/channel/shared_memory_channel.py:717]: update the comments

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 2]
```diff
@@ -646,6 +654,7 @@ def __init__(
         writer: Optional[ray.actor.ActorHandle],
         reader_and_node_list: List[Tuple["ray.actor.ActorHandle", str]],
         num_shm_buffers: int,
+        read_by_multi_output_node: bool,
```
Comment (kevin85421) [inline python/ray/experimental/channel/shared_memory_channel.py:657]: add doc string

File: python/ray/dag/compiled_dag_node.py [hunk 3]
```diff
@@ -737,7 +741,7 @@ def __init__(
         self._max_finished_execution_index: int = -1
         self._result_buffer: Dict[int, Any] = {}
 
-        def _get_creator_or_proxy_actor() -> "ray.actor.ActorHandle":
+        def _get_proxy_actor() -> "ray.actor.ActorHandle":
             """
             Get the creator actor or proxy actor handle of the DAG.
```
Comment (kevin85421) [inline python/ray/dag/compiled_dag_node.py:746]: update the comments

**Review by kevin85421 — COMMENTED** [review]
Overall, it looks good to me. My only concern is how much overhead will be introduced with the additional proxy actor? The overhead may be significant, especially since the main use case for creating a DAG from an actor is Ray Serve, which is sensitive to latency.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 2 above)
Comment (rkooo567) [reply python/ray/experimental/channel/shared_memory_channel.py:657]: done. and updated the var name to read_by_adag_driver

File: python/ray/dag/compiled_dag_node.py (hunk 3 above)
Comment (rkooo567) [reply python/ray/dag/compiled_dag_node.py:746]: actually I don't think it needs docstring. just removed

**Review by kevin85421 — APPROVED** [review]
LGTM. My only concern is how much overhead will be introduced with the additional proxy actor when the driver is an actor?

File: python/ray/dag/compiled_dag_node.py [hunk 4]
```diff
@@ -66,27 +66,32 @@ def do_allocate_channel(
     self,
     reader_and_node_list: List[Tuple["ray.actor.ActorHandle", str]],
     typ: ChannelOutputType,
+    read_by_adag_driver: bool,
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:69]: Is this equivalent to `is_adag_output_channel`? Should we use that name which seems easier to follow

Can we have a default value = False as most of the cases would be false

File: python/ray/dag/compiled_dag_node.py [hunk 5]
```diff
@@ -737,32 +742,7 @@ def __init__(
...
-            runtime_context = ray.get_runtime_context()
-            if runtime_context.worker.mode == ray.WORKER_MODE:
-                try:
-                    return ray.get_runtime_context().current_actor
-                except RuntimeError:
-                    raise NotImplementedError(
-                        "Compiled DAGs currently require the InputNode() to be the "
-                        "driver process or an actor method. Ray task is not supported."
-                    )
+        def _get_proxy_actor() -> "ray.actor.ActorHandle":
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:745]: nit: `_create_proxy_actor`

File: python/ray/dag/compiled_dag_node.py [hunk 6]
```diff
@@ -776,7 +756,7 @@ def _get_creator_or_proxy_actor() -> "ray.actor.ActorHandle":
                 )
             ).remote()
 
-        self._creator_or_proxy_actor = _get_creator_or_proxy_actor()
+        self._creator_or_proxy_actor = _get_proxy_actor()
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:759]: self._proxy_actor

File: python/ray/dag/class_node.py [hunk 7]
```diff
@@ -246,6 +246,9 @@ def _execute_impl(self, *args, **kwargs):
     def __str__(self) -> str:
         return get_dag_node_str(self, f"{self._method_name}()")
 
+    def __repr__(self) -> str:
```
Comment (ruisearch42) [inline python/ray/dag/class_node.py:249]: Question: why is this needed?
(__repr__ is not added in other DAGNode)

**Comment by rkooo567** [comment]
> LGTM. My only concern is how much overhead will be introduced with the additional proxy actor when the driver is an actor?

I believe this mechanism is not supposed to introduce any additional delay (it is how driver -> actor works). What's the potential overhead coming from?

File: python/ray/dag/compiled_dag_node.py (hunk 4 above)
Comment (rkooo567) [reply python/ray/dag/compiled_dag_node.py:69]: I think it is better not having default as it is kind of important flag that can cause hangs if set incorrectly. 

Regarding the name, I thought `is_adag_output_channel` exposes more implementation detail than this (that read by driver == output), and that was the original name before I changed it to this one. I don't have strong preference. Lmk if you want me to change.

**Review by ruisearch42 — APPROVED** [review]
Otherwise LGTM

File: python/ray/dag/compiled_dag_node.py (hunk 4 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:69]: The problem with `adag_driver` is it's not a well defined concept so you need to explain it everywhere and can be confused with ray driver. I don't have a good idea right now, so will leave it to you.

**Comment by kevin85421** [comment]
> I believe this mechanism is not supposed to introduce any additional delay (it is how driver -> actor works). What's the potential overhead coming from?

* Without this PR: actor 1 -> actor 2 -> actor 1
* With this PR: actor 1 -> actor 2 -> proxy actor -> actor 1

It needs to go through an additional proxy actor in this case.

**Comment by rkooo567** [comment]
@kevin85421 Does it affect the runtime performance? IIUC, we are not going through proxy actor when we read, so it should be okay.

**Comment by kevin85421** [comment]
> IIUC, we are not going through proxy actor when we read, so it should be okay.

Oh, I missed that part. Thanks!

File: python/ray/dag/class_node.py (hunk 7 above)
Comment (rkooo567) [reply python/ray/dag/class_node.py:249]: this was to make debugging easy when it was nested inside a list (in this case, repr is used).

File: python/ray/dag/compiled_dag_node.py (hunk 4 above)
Comment (rkooo567) [reply python/ray/dag/compiled_dag_node.py:69]: Hmm imo it is a reasonable wording, so unless there's strong pushback, I will probably just keep it
